pkg/storage/endpoint/gc_states.go

```diff
@@ -0,0 +1,556 @@
+		res.ExpiredAt = b.ExpirationTime.Unix()
+	}
+	return res
+}
+
+// IsExpired checks whether the GCBarrier is expired at the given time.
+func (b *GCBarrier) IsExpired(now time.Time) bool {
+	return b.ExpirationTime != nil && now.After(*b.ExpirationTime)
+}
+
+// String implements fmt.Stringer.
+func (b *GCBarrier) String() string {
+	expirationTime := "<nil>"
+	if b.ExpirationTime != nil {
+		expirationTime = b.ExpirationTime.String()
+	}
+	return fmt.Sprintf("GCBarrier { BarrierID: %s, BarrierTS: %d, ExpirationTime: %v }",
+		b.BarrierID, b.BarrierTS, expirationTime)
+}
+
+// GCStateStorage is the interface for providing the ability to store and retrieve GC state data.
+// The GC state data is not available to access via the GCStateStorage interface; instead, it should be
+// accessed by the GCStateProvider indirectly, which can be retrieved by calling GetGCStateProvider.
+type GCStateStorage interface {
+	// GetGCStateProvider returns an GCStateProvider for reading and writing GC state data.
+	GetGCStateProvider() GCStateProvider
+}
+
+// GetGCStateProvider returns an GCStateProvider for reading and writing GC state data.
+func (se *StorageEndpoint) GetGCStateProvider() GCStateProvider {
+	return newGCStateProvider(se)
+}
+
+// GCStateProvider is a stateless wrapper over StorageEndpoint that provides methods for reading/writing GC states.
+// It can be dangerous to misuse GC related operations. As an explicit wrapper, it hides the GC related methods away
+// from the StorageEndpoint type and the Storage interface, making it less likely to be misused unintentionally when
+// the Storage or StorageEndpoint is used in other context.
+type GCStateProvider struct {
+	storage *StorageEndpoint
+}
+
+// newGCStateProvider creates a new GCStateProvider.
+func newGCStateProvider(storage *StorageEndpoint) GCStateProvider {
+	return GCStateProvider{storage: storage}
+}
+
+// GCStateWriteBatch is the batch of write operations within a GCStateTransaction, which can be started by calling
+// GCStateProvider.RunInGCStateTransaction.
+type GCStateWriteBatch struct {
+	ops []kv.RawTxnOp
+}
+
+// LoadGCSafePoint loads current GC safe point of the given keyspaceID from storage.
+func (p GCStateProvider) LoadGCSafePoint(keyspaceID uint32) (uint64, error) {
+	if keyspaceID == constant.NullKeyspaceID {
+		return p.loadGlobalGCSafePoint()
+	}
+	return p.loadKeyspaceGCSafePoint(keyspaceID)
+}
+
+// loadGlobalGCSafePoint loads the GC safe point of the global GC.
+func (p GCStateProvider) loadGlobalGCSafePoint() (uint64, error) {
+	value, err := p.storage.Load(keypath.GCSafePointPath())
+	if err != nil || value == "" {
+		return 0, err
+	}
+	gcSafePoint, err := strconv.ParseUint(value, 16, 64)
+	if err != nil {
+		return 0, errs.ErrStrconvParseUint.Wrap(err).GenWithStackByArgs()
+	}
+	return gcSafePoint, nil
+}
+
+// keyspaceGCSafePoint is the data structure when writing
+type keyspaceGCSafePoint struct {
+	KeyspaceID uint32 `json:"keyspace_id"`
```

Comment:
What is the ID used for?

Reply:
Nothing but for keeping the data format compatible with the exisiting design of the "GC API V2". I forgot to finish the comments of this type. I'll fix it.